### ISSUE TITLE
MQTT Client Certificate Authentication

### ISSUE THREAD

**Issue by Daniel-dev22** [bkbilly/lnxlink#87]
### Describe the solution you'd like

MQTT Client Certificate Authentication can also be referred to as MQTT MTLS.

For this to work additional Mqtt parameters need to be added to the code, client certificate and client key in addition to the certificate authority certificate.

This is how pahomqtt does it. I just got this to work in another project.

https://github.com/jgyates/genmon/issues/1006#issuecomment-1882932513

**Comment by bkbilly** [comment]
When you setup LNXlink, it asks you if you want to use TLS.
You can manually edit your yaml config file so that you can use TLS connection for MQTT.
This looks like so:
```yaml
mqtt:
  prefix: 'lnxlink'
  clientId: 'DESKTOP-Linux'
  server: '192.168.1.1'
  port: 1883
  auth:
    user: 'user'
    pass: 'pass'
    tls: true
  discovery:
    enabled: true
  lwt:
    enabled: true
    qos: 1
    retain: false
    connectMsg: 'ON'
    disconnectMsg: 'OFF'
```

You can see how I use it here: https://github.com/bkbilly/lnxlink/blob/master/lnxlink/__main__.py#L163

This feature was introduced on 2023.9.0 version.

**Comment by bkbilly** [comment]
I could add this option for providing a key file, but I am going to need some help from you because I don't have an MQTT broker with a TLS verification to test it.

You should install LNXlink as a developer and I would like to know if these changes work for your installation:
Replace [this line](https://github.com/bkbilly/lnxlink/blob/master/lnxlink/__main__.py#L164) with this code:
```python
            self.client.tls_set(certfile=None, keyfile="certificate.pem", cert_reqs=ssl.CERT_REQUIRED)
```
You could delete this line to see if it works:
```python
            self.client.tls_insecure_set(True)
```

If these changes work, I can start working on adding the `certificate.pem` file in the yaml configuration file.

**Comment by bkbilly** [comment]
I've added options for adding the files for certificate and key.
Let me know if it works for you.

**Comment by bkbilly** [comment]
I've added the ca_cert option.

**Comment by bkbilly** [comment]
Oh, you are right. I made a mistake of how these files are read.
I've updated it read them as flat.

**Comment by bkbilly** [comment]
I would recommend to add the full path of the files because most likely it looks at the `__main__.py` location.

**Comment by bkbilly** [comment]
I've made some changes, which hopefully, should make this work.
For the certificates, you need to use absolute paths and be able to read them by the user who runs the LNXlink app.

**Comment by bkbilly** [comment]
It seems that you are not running LNXlink as a developer, so the changes I've made, won't work for you.
You can wait for the next release, or test it by uninstalling the current version and follow the instructions to install as a developer.

**Comment by bkbilly** [comment]
Oh, I see how it works.
I've been trying to connect to my broker I created and I couldn't, but adding the username/password, it works perfectly.

I've pushed an update that includes the username/password check you suggested.

**Comment by bkbilly** [comment]
The tls_insecure_set is set to False by default.
I've tried it on my system with a self signed certificate which won't connect unless I set it to True.

Thanks for helping out improving the security of LNXlink!
I've also learned a thing or two in the process.

**Comment by bkbilly** [comment]
I released the 2024.2.0 version that includes this fix.
